src/compiler/checker.ts

```diff
@@ -6052,6 +6066,28 @@ namespace ts {
                 }
                 return Ternary.True;
             }
+
+            function constructorRelatedTo(sourceSignature: Signature, targetSignature: Signature, reportErrors: boolean) {
+                if (sourceSignature && targetSignature && sourceSignature.declaration && targetSignature.declaration) {
+                    // A public, protected and private signature is assignable to a private signature.
+                    // A public and protected signature is assignable to a protected signature.
+                    // And only a public signature is assignable to public signature.
+                    const sourceAccessibility = sourceSignature.declaration.flags & (NodeFlags.Private | NodeFlags.Protected);
+                    const targetAccessibility = targetSignature.declaration.flags & (NodeFlags.Private | NodeFlags.Protected);
+
+                    const isRelated = targetAccessibility === NodeFlags.Private
+                    || (targetAccessibility === NodeFlags.Protected && sourceAccessibility !== NodeFlags.Private)

```

Comment:
It'd just be clearer if you formatted this as

// A public, protected and private signature is assignable to a private signature.
if (targetAccessibility === NodeFlags.Private) {
    return true;
}

// A public and protected signature is assignable to a protected signature.
if (targetAccessibility === NodeFlags.Protected && sourceAccessibility !== NodeFlags.Private) {
    return true;
}

// Only a public signature is assignable to public signature.
if (targetAccessibility !== NodeFlags.Protected && !sourceAccessibility) {
    return true;
}

if (reportErrors) {
    reportError(Diagnostics.Cannot_assign_a_0_constructor_type_to_a_1_constructor_type, visibilityToString(sourceAccessibility), visibilityToString(targetAccessibility));
}
return false;
```